File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/SeedDerivation.java
package com.tongsr.kaleido.gradle;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.Normalizer;
import java.util.HexFormat;

final class SeedDerivation {
    static final String SEED_POLICY_VERSION = "kaleido-seed-v1";

    private SeedDerivation() {}

    static String fingerprint(String rawSeed) {
        return fingerprint(rawSeed, "<unknown>", "<unknown>");
    }

    static String fingerprint(String rawSeed, String projectPath, String variantName) {
        if (rawSeed == null || rawSeed.isBlank()) {
            throw new KaleidoDiagnostic(
                    "KLD-CONFIG-001",
                    projectPath,
                    variantName,
                    "adoption-plan",
                    "kaleido.seed",
                    "seed",
                    "Seed Provider resolved to a blank value",
                    "Provide a nonblank deterministic seed").failure();
        }
        return sha256(Normalizer.normalize(rawSeed, Normalizer.Form.NFC));
    }

    static String defaultFingerprint(String applicationId, String variantIdentity) {
        var material = String.join("\u0000",
                SEED_POLICY_VERSION,
                Normalizer.normalize(applicationId, Normalizer.Form.NFC),
                Normalizer.normalize(variantIdentity, Normalizer.Form.NFC));
        return sha256(material);
    }

    static String derive(String rootFingerprint, String domain, String variantIdentity) {
        var material = String.join("\u0000",
                "kaleido-domain-seed-v1",
                domain,
                rootFingerprint,
                variantIdentity);
        return sha256(material);
    }

    static org.gradle.api.GradleException missingProviderValue(
            String projectPath, String variantName) {
        return new KaleidoDiagnostic(
                "KLD-CONFIG-001",
                projectPath,
                variantName,
                "adoption-plan",
                "kaleido.seed",
                "seed",
                "Explicit seed Provider resolved no value",
                "Make the configured seed Provider present for every Release build").failure();
    }

    private static String sha256(String normalizedValue) {
        try {
            var digest = MessageDigest.getInstance("SHA-256")
                    .digest(normalizedValue.getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(digest);
        } catch (NoSuchAlgorithmException impossible) {
            throw new IllegalStateException("SHA-256 is unavailable", impossible);
        }
    }

}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/SeedDerivation.kt
package com.tongsr.kaleido.gradle

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.text.Normalizer
import java.util.HexFormat
import org.gradle.api.GradleException

internal object SeedDerivation {
    const val SEED_POLICY_VERSION: String = "kaleido-seed-v1"

    @JvmStatic
    fun fingerprint(rawSeed: String): String =
        fingerprint(rawSeed, "<unknown>", "<unknown>")

    @JvmStatic
    fun fingerprint(rawSeed: String?, projectPath: String, variantName: String): String {
        if (rawSeed.isNullOrBlank()) {
            throw KaleidoDiagnostic(
                "KLD-CONFIG-001",
                projectPath,
                variantName,
                "adoption-plan",
                "kaleido.seed",
                "seed",
                "Seed Provider resolved to a blank value",
                "Provide a nonblank deterministic seed",
            ).failure()
        }
        return sha256(Normalizer.normalize(rawSeed, Normalizer.Form.NFC))
    }

    @JvmStatic
    fun defaultFingerprint(applicationId: String, variantIdentity: String): String {
        val material = listOf(
            SEED_POLICY_VERSION,
            Normalizer.normalize(applicationId, Normalizer.Form.NFC),
            Normalizer.normalize(variantIdentity, Normalizer.Form.NFC),
        ).joinToString("\u0000")
        return sha256(material)
    }

    @JvmStatic
    fun derive(rootFingerprint: String, domain: String, variantIdentity: String): String {
        val material = listOf(
            "kaleido-domain-seed-v1",
            domain,
            rootFingerprint,
            variantIdentity,
        ).joinToString("\u0000")
        return sha256(material)
    }

    fun missingProviderValue(projectPath: String, variantName: String): GradleException =
        KaleidoDiagnostic(
            "KLD-CONFIG-001",
            projectPath,
            variantName,
            "adoption-plan",
            "kaleido.seed",
            "seed",
            "Explicit seed Provider resolved no value",
            "Make the configured seed Provider present for every Release build",
        ).failure()

    private fun sha256(normalizedValue: String): String {
        return try {
            val digest = MessageDigest.getInstance("SHA-256")
                .digest(normalizedValue.toByteArray(StandardCharsets.UTF_8))
            HexFormat.of().formatHex(digest)
        } catch (impossible: NoSuchAlgorithmException) {
            throw IllegalStateException("SHA-256 is unavailable", impossible)
        }
    }
}
